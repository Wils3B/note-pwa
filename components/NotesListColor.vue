<template>
  <div class="note-list-color">
    <v-expansion-panels multiple>
      <v-expansion-panel v-for="(el, index) of sorted" :key="index">
        <v-expansion-panel-header expand-icon="expand_more">
          {{ el.color }} ( {{ el.notes.length }} )
        </v-expansion-panel-header>
        <v-expansion-panel-content color="brown">
          <app-note-list
            :notes="el.notes"
            :sizes="{ cols: 12, sm: 6, md: 4, lg: 3, xl: 2 }"
          />
        </v-expansion-panel-content>
      </v-expansion-panel>
    </v-expansion-panels>
  </div>
</template>

<script>
import AppNoteList from '~/components/NotesList.vue'

export default {
  name: 'NoteListColor',
  components: {
    AppNoteList
  },
  props: {
    notes: {
      type: Array,
      default: () => []
    }
  },
  computed: {
    sorted() {
      if (this.notes.length === 0) return []
      const colors = this.$store.state.colors
      const values = colors
        .map((color) => {
          return {
            color,
            notes: this.notes.filter((note) => note.color === color)
          }
        })
        .filter((el) => el.notes.length !== 0)
        .sort((a, b) => b.notes.length - a.notes.length)
      return values
    }
  }
}
</script>

<style lang="scss">
.note-list-color {
  .v-expansion-panel-content__wrap {
    padding-top: 8px;
    padding-bottom: 8px;
  }
}
</style>
