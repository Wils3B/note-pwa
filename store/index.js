export const state = () => {
  return {
    colors: ['white', 'yellow', 'purple', 'green', 'red', 'blue', 'brown'],
  }
}
