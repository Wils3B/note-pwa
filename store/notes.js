import DB from '~/model/db'
import Note from '~/model/note'

export const state = () => {
  const notes = []
  return {
    notes,
  }
}

export const mutations = {
  saveNote(currentState, note) {
    currentState.notes.push(note)
    DB.addNotes(note)
  },
  deleteNote(currentState, id) {
    const note = currentState.notes.filter((note) => note.id === id)[0]
    const index = currentState.notes.indexOf(note)
    currentState.notes.splice(index, 1)
    DB.deleteNotes(id)
  },
  editNote(currentState, note) {
    const index = currentState.notes.indexOf(
      getters.noteById(currentState)(note.id)
    )
    const currentNote = currentState.notes[index]
    currentNote.color = note.color
    Note.setContent(currentNote, note.content)
    currentState.notes.splice(index, 1)
    currentState.notes.push(currentNote)
    DB.putNotes(note)
  },
  setAllNotes(currentState, notes) {
    currentState.notes = notes
  },
}

export const getters = {
  noteById(currentState) {
    window.notes = currentState.notes
    return (id) => currentState.notes.filter((note) => note.id === id)[0]
  },
}

export const actions = {
  async fetchAllNotes(context) {
    if (!DB.connected) await DB.connect()
    let notes
    if (DB.connected) {
      notes = await DB.getAllNotes()
      context.commit(
        'setAllNotes',
        notes.data.sort((a, b) => (a.modifiedAt < b.modifiedAt ? -1 : 1))
      )
    }
  },
}
