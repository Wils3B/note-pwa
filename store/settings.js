export const state = () => {
  const loadedSettings = loadSettings()
  return {
    selecteds: {
      sortType: loadedSettings.sortType || 'date'
    },
    groups: {
      sortTypes: ['date', 'color']
    }
  }
}

export const mutations = {
  changeSortType(currentState, type) {
    currentState.selecteds.sortType = type
    storeSettings(currentState.selecteds)
  }
}

function storeSettings(settings) {
  localStorage.setItem('settings', JSON.stringify(settings))
}

function loadSettings() {
  const settingsJSON = localStorage.getItem('settings')
  let settingsObj = {}
  try {
    settingsObj = JSON.parse(settingsJSON)
  } catch (e) {}
  return settingsObj || {}
}
