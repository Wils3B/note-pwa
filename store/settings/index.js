import { loadSettings, storeSettings } from './functions'

export const state = () => {
  const loadedSettings = loadSettings()
  return {
    selecteds: {
      sortType: loadedSettings.sortType || 'date',
      defaultColor: loadedSettings.defaultColor || 'white',
      appFont: loadedSettings.appFont || 'Amita',
      autosave: loadedSettings.autosave || '2min',
    },
    groups: {
      sortTypes: ['date', 'color'],
      fonts: ['Kalam', 'Amita', 'Roboto'],
      autosave: ['off', '10s', '20s', '30s', '2min', '5min'],
    },
  }
}

export const mutations = {
  changeSortType(currentState, type) {
    currentState.selecteds.sortType = type
    storeSettings(currentState.selecteds)
  },
  changeDefaultColor(currentState, color) {
    currentState.selecteds.defaultColor = color
    storeSettings(currentState.selecteds)
  },
  changeAppFont(currentState, font) {
    currentState.selecteds.appFont = font
    storeSettings(currentState.selecteds)
  },
  changeAutosave(currentState, autosave) {
    currentState.selecteds.autosave = autosave
    storeSettings(currentState.selecteds)
  },
}

export const getters = {
  autosaveInMillis(currentState) {
    const setting = currentState.selecteds.autosave
    if (/^\d+s$/.test(setting)) {
      return 1000 * Number(setting.substring(0, setting.length - 1))
    } else if (/^\d+min$/.test(setting)) {
      return 1000 * 60 * Number(setting.substring(0, setting.length - 3))
    }
    return 0
  },
}
