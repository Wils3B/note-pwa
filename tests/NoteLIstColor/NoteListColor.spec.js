import { mount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import { localInstance } from '../common'
import router from '../router'
import store from '../store'
import notes from '../notes'
import NotesListColor from '~/components/NotesListColor.vue'

const localVue = localInstance()

describe('Note List Color', () => {
  const wrapper = mount(NotesListColor, {
    localVue,
    router,
    store,
    vuetify: new Vuetify(),
    propsData: {
      notes,
    },
  })

  test('it should be a vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy()
  })

  test('Have a correct render', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('note grouping', () => {
    const groups = wrapper.findAll('.v-expansion-panel')

    test('it should have 6 group by color', () => {
      expect(groups.length).toBe(6)
    })

    test('each group should be in correct order and have correct number of notes', () => {
      const colors = wrapper.vm.$store.state.colors
        .map((color) => {
          let count = 0
          notes.forEach((n) => (count += n.color === color ? 1 : 0))
          return { color, count }
        })
        .sort((a, b) => a.count > b.count)
      colors.splice(colors.find((color) => color.count === 0))

      colors.forEach((color, index) => {
        expect(groups.at(index).text()).toBe(`${color.color} ( ${color.count} )`)
      })
    })

    test('it should not have any group if no note is passed', () => {
      const localWrapper = mount(NotesListColor, {
        localVue,
        router,
        store,
      })
      expect(localWrapper.findAll('.v-expansion-panel').length).toBe(0)
    })
  })
})
