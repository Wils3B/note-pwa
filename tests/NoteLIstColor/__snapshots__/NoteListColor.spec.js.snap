// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note List Color Have a correct render 1`] = `
"<div class=\\"note-list-color\\">
  <div class=\\"v-item-group theme--light v-expansion-panels\\">
    <div aria-expanded=\\"false\\" class=\\"v-expansion-panel\\"><button type=\\"button\\" class=\\"v-expansion-panel-header\\">
        yellow ( 4 )
        <transition-stub name=\\"fade-transition\\">
          <div class=\\"v-expansion-panel-header__icon\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">expand_more</i></div>
        </transition-stub></button>
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
    <div aria-expanded=\\"false\\" class=\\"v-expansion-panel\\"><button type=\\"button\\" class=\\"v-expansion-panel-header\\">
        white ( 2 )
        <transition-stub name=\\"fade-transition\\">
          <div class=\\"v-expansion-panel-header__icon\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">expand_more</i></div>
        </transition-stub></button>
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
    <div aria-expanded=\\"false\\" class=\\"v-expansion-panel\\"><button type=\\"button\\" class=\\"v-expansion-panel-header\\">
        green ( 2 )
        <transition-stub name=\\"fade-transition\\">
          <div class=\\"v-expansion-panel-header__icon\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">expand_more</i></div>
        </transition-stub></button>
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
    <div aria-expanded=\\"false\\" class=\\"v-expansion-panel\\"><button type=\\"button\\" class=\\"v-expansion-panel-header\\">
        blue ( 2 )
        <transition-stub name=\\"fade-transition\\">
          <div class=\\"v-expansion-panel-header__icon\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">expand_more</i></div>
        </transition-stub></button>
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
    <div aria-expanded=\\"false\\" class=\\"v-expansion-panel\\"><button type=\\"button\\" class=\\"v-expansion-panel-header\\">
        purple ( 1 )
        <transition-stub name=\\"fade-transition\\">
          <div class=\\"v-expansion-panel-header__icon\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">expand_more</i></div>
        </transition-stub></button>
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
    <div aria-expanded=\\"false\\" class=\\"v-expansion-panel\\"><button type=\\"button\\" class=\\"v-expansion-panel-header\\">
        red ( 1 )
        <transition-stub name=\\"fade-transition\\">
          <div class=\\"v-expansion-panel-header__icon\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate material-icons theme--light\\">expand_more</i></div>
        </transition-stub></button>
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
  </div>
</div>"
`;
